src/main/java/io/jenkins/plugins/coverage/adapter/DListingReportAdapter.java

```diff
@@ -0,0 +1,195 @@
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public final class DListingReportAdapter extends CoverageReportAdapter {
+
+    private static final String UNCOVERED = "0000000";
+    private static final String NON_EXECUTABLE_CODE = "";
+
+    @DataBoundConstructor
+    public DListingReportAdapter(String path) {
+
+        super(path == null || path.equals("") ? "*.lst" : path);
+        setMergeToOneReport(true);
+    }
+
+    @Override
+    public Document convert(File source) throws CoverageException {
+        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
+        try {
+            factory.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
+        } catch (ParserConfigurationException e) {
+            e.printStackTrace();
+        }
+
+        factory.setNamespaceAware(true);
+        DocumentBuilder builder;
+        try {
+            builder = factory.newDocumentBuilder();
+        } catch (ParserConfigurationException e) {
+            throw new CoverageException(e);
+        }
+
+        Document document = builder.newDocument();
+        Element rootElement = document.createElement("report");
+        rootElement.setAttribute("name", "D Listing Coverage");
+        document.appendChild(rootElement);
+
+        Element fileElement = document.createElement("file");
+
+        try (BufferedReader br = Files.newBufferedReader(Paths.get(source.getPath()))) {
+            List<String> list = br.lines().collect(Collectors.toList());
+
+            if (list.size() == 0) {
+                return document;
+            }
+
+            String lastLine = list.get(list.size() - 1);
+            String moduleFileName = extractModuleFileName(lastLine);
+
+            if (moduleFileName == null) {
+                return document;
+            }
+
+            fileElement.setAttribute("name", moduleFileName);
+
+            int i = 0;
+            for (String line : list) {
+                String[] parts = line.split("\\|");
+                if (parts.length > 1) {
+                    String lineStatistic = parts[0].trim();
+
+                    if (!lineStatistic.equals(NON_EXECUTABLE_CODE)) {
+                        Element lineElement = document.createElement("line");
+                        lineElement.setAttribute("number", Integer.toString(i + 1));
+
+                        String hits = (lineStatistic.equals(UNCOVERED)) ? "0" : lineStatistic.trim();
+                        lineElement.setAttribute("hits", hits);
+                        fileElement.appendChild(lineElement);
+                    }
+                }
+                i++;
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+
+        rootElement.appendChild(fileElement);
+        return document;
+    }
+
+    private String extractModuleFileName(String line) {
+        Pattern patternIsCovered = Pattern.compile("(.*) is \\d+% covered");
```

Comment:
It is better to change them to static field so that we don't have to compile them every time we call this method.

Reply:
yes, will change